### PR TITLE
cli: reap spawned wireguard child

### LINKED ISSUE #132: `rosenpass` not reaping `wireguard` children
rosenpass 0.2.0 on nixos (`nix-info`: `system: "x86_64-linux", multi-user?: yes, version: nix-env (Nix) 2.17.0, nixpkgs: /nix/store/cz973qsjldbw4x7fx0rwcvhr9k645vz2-source`)

```
root       48686  0.0  0.0   9304  5572 ?        SL   09:43   0:00  \_ rosenpass exchange-config /nix/store/jbwz87z7cppqj466fkb3i33dcx53rx4a-rosenpass-config
root       49564  0.0  0.0      0     0 ?        Z    09:43   0:00      \_ [.wg-wrapped] <defunct>
root       51547  0.0  0.0      0     0 ?        Z    09:45   0:00      \_ [.wg-wrapped] <defunct>
root       53545  0.0  0.0      0     0 ?        Z    09:47   0:00      \_ [.wg-wrapped] <defunct>
root       55322  0.0  0.0      0     0 ?        Z    09:49   0:00      \_ [.wg-wrapped] <defunct>
root       57056  0.0  0.0      0     0 ?        Z    09:52   0:00      \_ [.wg-wrapped] <defunct>
root       58795  0.0  0.0      0     0 ?        Z    09:54   0:00      \_ [.wg-wrapped] <defunct>
root       60579  0.0  0.0      0     0 ?        Z    09:56   0:00      \_ [.wg-wrapped] <defunct>
root       62813  0.0  0.0      0     0 ?        Z    09:58   0:00      \_ [.wg-wrapped] <defunct>
root       65028  0.0  0.0      0     0 ?        Z    10:00   0:00      \_ [.wg-wrapped] <defunct>
root       67174  0.0  0.0      0     0 ?        Z    10:02   0:00      \_ [.wg-wrapped] <defunct>
root       68821  0.0  0.0      0     0 ?        Z    10:05   0:00      \_ [.wg-wrapped] <defunct>
root       70682  0.0  0.0      0     0 ?        Z    10:07   0:00      \_ [.wg-wrapped] <defunct>
root       72456  0.0  0.0      0     0 ?        Z    10:09   0:00      \_ [.wg-wrapped] <defunct>
root       74073  0.0  0.0      0     0 ?        Z    10:11   0:00      \_ [.wg-wrapped] <defunct>
root       75929  0.0  0.0      0     0 ?        Z    10:13   0:00      \_ [.wg-wrapped] <defunct>
root       77574  0.0  0.0      0     0 ?        Z    10:15   0:00      \_ [.wg-wrapped] <defunct>
root       80236  0.0  0.0      0     0 ?        Z    10:18   0:00      \_ [.wg-wrapped] <defunct>
root       82174  0.0  0.0      0     0 ?        Z    10:20   0:00      \_ [.wg-wrapped] <defunct>
root       84515  0.0  0.0      0     0 ?        Z    10:22   0:00      \_ [.wg-wrapped] <defunct>
root       86544  0.0  0.0      0     0 ?        Z    10:24   0:00      \_ [.wg-wrapped] <defunct>
root       88917  0.0  0.0      0     0 ?        Z    10:26   0:00      \_ [.wg-wrapped] <defunct>
root       91170  0.0  0.0      0     0 ?        Z    10:28   0:00      \_ [.wg-wrapped] <defunct>
root       93390  0.0  0.0      0     0 ?        Z    10:31   0:00      \_ [.wg-wrapped] <defunct>
root       95783  0.0  0.0      0     0 ?        Z    10:33   0:00      \_ [.wg-wrapped] <defunct>
root       98073  0.0  0.0      0     0 ?        Z    10:35   0:00      \_ [.wg-wrapped] <defunct>
root      100047  0.0  0.0      0     0 ?        Z    10:37   0:00      \_ [.wg-wrapped] <defunct>
root      102286  0.0  0.0      0     0 ?        Z    10:39   0:00      \_ [.wg-wrapped] <defunct>
root      104222  0.0  0.0      0     0 ?        Z    10:41   0:00      \_ [.wg-wrapped] <defunct>
root      106166  0.0  0.0      0     0 ?        Z    10:44   0:00      \_ [.wg-wrapped] <defunct>
root      108857  0.0  0.0      0     0 ?        Z    10:46   0:00      \_ [.wg-wrapped] <defunct>
root      110657  0.0  0.0      0     0 ?        Z    10:48   0:00      \_ [.wg-wrapped] <defunct>
root      112578  0.0  0.0      0     0 ?        Z    10:50   0:00      \_ [.wg-wrapped] <defunct>
root      114443  0.0  0.0      0     0 ?        Z    10:52   0:00      \_ [.wg-wrapped] <defunct>
root      116235  0.0  0.0      0     0 ?        Z    10:54   0:00      \_ [.wg-wrapped] <defunct>
root      117977  0.0  0.0      0     0 ?        Z    10:57   0:00      \_ [.wg-wrapped] <defunct>
root      119734  0.0  0.0      0     0 ?        Z    10:59   0:00      \_ [.wg-wrapped] <defunct>
root      121449  0.0  0.0      0     0 ?        Z    11:01   0:00      \_ [.wg-wrapped] <defunct>
root      123256  0.0  0.0      0     0 ?        Z    11:03   0:00      \_ [.wg-wrapped] <defunct>
root      124916  0.0  0.0      0     0 ?        Z    11:05   0:00      \_ [.wg-wrapped] <defunct>
```

### PR DESCRIPTION
**Pull request by wahjava**
Fixes #132

### CODE REVIEW

**Comment by emilengler** [comment]
```diff
diff --git a/src/app_server.rs b/src/app_server.rs
index 39cd312..1c7f094 100644
--- a/src/app_server.rs
+++ b/src/app_server.rs
@@ -1,6 +1,7 @@
 use anyhow::bail;

 use anyhow::Result;
+use log::debug;
 use log::{error, info, warn};
 use mio::Interest;
 use mio::Token;
@@ -633,7 +634,13 @@ impl AppServer {
                 .args(&owg.extra_params)
                 .spawn()?;
             b64_writer(child.stdin.take().unwrap()).write_all(key.secret())?;
-            child.wait()?;
+            let status = child.wait()?;
+
+            if status.success() {
+                debug!("successfully passed psk to wg");
+            } else {
+                error!("could not pass psk to wg {:?}", status);
+            }
         }

         Ok(())
```
So far, this looks good to me. However, when we already wait for the child to exit, we can also check its return value to see if things worked out as we intended to. Kinda unsure if we should "only" log the error or panic instead. Any thoughts @wucke13?

File: src/app_server.rs [hunk 1]
```diff
@@ -632,7 +632,14 @@ impl AppServer {
                 .stdin(Stdio::piped())
                 .args(&owg.extra_params)
                 .spawn()?;
-            b64_writer(child.stdin.unwrap()).write_all(key.secret())?;
+            b64_writer(child.stdin.take().unwrap()).write_all(key.secret())?;
+            let status = child.wait()?;
```
Comment (wucke13) [inline src/app_server.rs:636]: This is dangerous, now timing of our event loop is distracted by the `wg` command.

**Comment by wucke13** [comment]
If we must wait and check for the output, this should happen in another thread. Is there a way to spawn the `wg` without causing zombies?

**Comment by koraa** [comment]
@wahjava Would you like a tip for this contribution through our [bounty provider](https://algora.io)?

**Comment by emilengler** [comment]
@wucke13 Your thoughts on using a pidfd instead? It can be fetched asynchronously to see if a process is dead.

**Comment by koraa** [comment]
> When we already wait for the child to exit, we can also check its return value to see if things worked out as we intended to. Kinda unsure if we should "only" log the error or panic instead.

We should log the error: When we crash the rosenpass process, we should always make sure that the current wireguard PSK is erased; our chances of accomplishing that are higher when we try again.



> > @wucke13 Your thoughts on using a pidfd instead? It can be fetched asynchronously to see if a process is dead.
> 
> One would still need to `wait()` to reap the zombie process, if I'm not mistaken. And `pidfd` is Linux-only.

Agreed, I think using a separate thread is good enough. LGTTM!

**Review by koraa — APPROVED** [review]

**Pull request merged**